### PR TITLE
API Specification Draft 0.1

### PR DESCRIPTION
**Pull request by mmisztal1980**
The existing discussion has been converted to an in-progress API v1 draft.
- some english language mistakes have been corrected, and wording has been changed
- some concepts have been altered due to inconsistencies found in the original discussion

Initial draft contains:
- device registration api
- device data ingress api
- data egress apis

Topics not covered:
- security
- MQTT
- Geo-spatial queries

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -9,369 +59,28 @@
...
-		},
-		"latitude": 40.713000,
-		"longitude": -74.013500,
-		"elevation": 546.0
-	}
-}
-```
+|Symbol       |Name            |Unit   |
+--------------|----------------|-------|
+CO            |Carbon Monoxide |[mg/m3]
```
Comment (MaciejGad) [inline README.md]: should be `m³` instead of `m3`

Comment (mmisztal1980) [inline README.md]: fixed

File: api/data-egress/REST/readme.md [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+```json
+{
+	"value": "<list of strings containing city names>",
+	"correlationId": "string",
+	"paging": {
+		"index": "int",
+		"size": "int",
+		"pages": "int"
+	},
+	"hasError" : "boolean",
```
Comment (ochococo) [inline api/data-egress/REST/readme.md]: `"<boolean>"` ?

File: api/data-egress/REST/readme.md [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+- All egress API operations, related to device information retrieval shall start with a `/devices` url segment.
+- All egress API operations, related to data retrieval shall start with a `/data` url segment.
+- All egress API operations shall use the `GET` HTTP verb.
+
+## 4.4.1 General Location and Count retrieval
+
+### 4.4.1.1 Cities containing sensors
+
+Return a list of a country's cities, containing at least 1 sensor.
+- `GET {air quality api base address}/v1/devices/{country}/cities`
```
Comment (ochococo) [inline api/data-egress/REST/readme.md]: `{country}` - a number? ISO code? UUID?

File: api/data-egress/REST/readme.md [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+
+```json
+{
+	"value": "<list of strings containing city names>",
+	"correlationId": "string",
```
Comment (mmisztal1980) [inline api/data-egress/REST/readme.md]: What is unclear?

File: api/data-egress/REST/readme.md (hunk 2 above)
Comment (mmisztal1980) [inline api/data-egress/REST/readme.md]: Right

File: api/data-egress/REST/readme.md (hunk 3 above)
Comment (mmisztal1980) [inline api/data-egress/REST/readme.md]: ISO code

Comment (ochococo) [inline api/data-egress/REST/readme.md]: Should be defined somewhere.

File: api/data-egress/REST/readme.md (hunk 4 above)
Comment (ochococo) [inline api/data-egress/REST/readme.md]: `correlationId` - what is it, what for, why should one expect it?

File: api/data-egress/REST/readme.md [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+500           | Internal server error (Unhandled exception while processing the request)
+
+*Example*
+
+- `GET {air quality api base address}/v1/devices/pl/cities`
+
+```json
+{
+	"value": [
+        "katowice"
```
Comment (ochococo) [inline api/data-egress/REST/readme.md]: Nah - you can expect collisions - pola, póla, poła, polą etc. maybe `123-warsaw` or just `123`.

File: api/data-egress/REST/readme.md (hunk 4 above)
Comment (mmisztal1980) [inline api/data-egress/REST/readme.md]: correlationId is a UUID assigned to each request, used to track requests in logs, across service calls. 
I think [this](http://theburningmonk.com/2015/05/a-consistent-approach-to-track-correlation-ids-through-microservices/) article eplains the concept pretty well.

Comment (ochococo) [inline api/data-egress/REST/readme.md]: makes perfect sense now 👍

File: api/data-egress/REST/readme.md (hunk 5 above)
Comment (mmisztal1980) [inline api/data-egress/REST/readme.md]: perhaps we should enforce ASCII characters only?

Comment (ochococo) [inline api/data-egress/REST/readme.md]: Well THAT is the problem `Pola` and `Poła` == `pola`. Collision.

Comment (mmisztal1980) [inline api/data-egress/REST/readme.md]: Is this better for you?
```json
{
    "value": [
        { "id": 123, "name": "katowice" }
    ]
}
```

Comment (mmisztal1980) [inline api/data-egress/REST/readme.md]: I'm having another look at entire 4.4.1. I think I can design it a bit "better"

File: api/data-egress/REST/readme.md (hunk 2 above)
Comment (mmisztal1980) [inline api/data-egress/REST/readme.md]: fixed everywhere (I think). All types should now be enclosed within "<typeName>"

File: api/data-egress/REST/readme.md (hunk 3 above)
Comment (mmisztal1980) [inline api/data-egress/REST/readme.md]: added reference to ISO-3166-1

File: api/data-egress/REST/readme.md (hunk 5 above)
Comment (mmisztal1980) [inline api/data-egress/REST/readme.md]: I've had another go at the `/devices` api specification. please have another look.

**Comment by mmisztal1980** [comment]
merged

**Pull request merged**